core/Pos/Core/Vss.hs

```diff
@@ -71,22 +86,75 @@ checkCertSign VssCertificate {..} =
 getCertId :: VssCertificate -> StakeholderId
 getCertId = addressHash . vcSigningKey
 
--- | Safe constructor of 'VssCertificatesMap'. TODO: wrap into newtype.
-mkVssCertificatesMap :: [VssCertificate] -> VssCertificatesMap
-mkVssCertificatesMap = HM.fromList . map toCertPair
+-- | Safe constructor of 'VssCertificatesMap'. It doesn't allow certificates
+-- with duplicate signing keys or with duplicate 'vcVssKey's.
+mkVssCertificatesMap
+    :: MonadFail m
+    => [VssCertificate] -> m VssCertificatesMap
+mkVssCertificatesMap certs = do
+    unless (allDistinct (map vcSigningKey certs)) $
+        fail "mkVssCertificatesMap: two certs have the same signing key"
+    unless (allDistinct (map vcVssKey certs)) $
+        fail "mkVssCertificatesMap: two certs have the same VSS key"
+    pure $ UnsafeVssCertificatesMap (HM.fromList (map toCertPair certs))
```

Comment:
I would write `pure $ mkVssCertificatesMapLossy certs` here to slightly decrease code duplication (no need to define `toCertPair` in two places).

Reply:
I don't want to do this because `nubOrdOn` isn't cheap and here we can avoid it.

Reply:
Okay, I will define `toCertPair` as a top-level function because it can be used in three places.

Reply:
Deal.